cpp/src/arrow/dataset/discovery.cc

```diff
@@ -146,18 +170,18 @@ Result<DataSourcePtr> FileSystemDataSourceDiscovery::Finish() {
 
       if (segments.size() > 0) {
         auto segment_index = static_cast<int>(segments.size()) - 1;
-        auto maybe_partition = partition_scheme_->Parse(segments.back(), segment_index);
+        auto maybe_partition = partition_scheme->Parse(segments.back(), segment_index);
 
         partitions[ref.i] = std::move(maybe_partition).ValueOr(scalar(true));
       }
     }
     return Status::OK();
   };
 
-  RETURN_NOT_OK(forest.Visit(apply_partition_scheme));
+  RETURN_NOT_OK(forest_.Visit(apply_partition_scheme));
```

Comment:
I'm not sure how you want to deal here, either nullptr or NoopScheme. I tend to prefer nullptr, but you don't.

Reply:
DefaultPartitionScheme is a noop and it's the old default for the partition_scheme field. I'll just re-add that